Add DMZ Examples, templates 411, 412, and 413

tracsman opened the pull request:
### Contributing guide

https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md
### Changelog
- Adding Examples 1, 2, and 3 from the Azure.com Network Security whitepaper
- Initial upload
### Description of the change

I'm updating the Network Security Best Practice whitepaper on Azure.com to include ARM templates. These are said templates with README.md, Deploy json, parameters json, and metadata json files.

[https://azure.microsoft.com/documentation/articles/best-practices-network-security/](https://azure.microsoft.com/documentation/articles/best-practices-network-security/)

azurecla commented:
Hi __@tracsman__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (jonor). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

singhkay commented on 411-dmz-example-1/azuredeploy.json:
```diff
@@ -0,0 +1,521 @@
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "newStorageAccountName": {
```
See #944 for storage guidance

tracsman replied:
Removed parameter and added a new variable:
"StorageAccountName": "[concat(uniquestring(resourceGroup().id), 'dmz1')]",

singhkay commented on 411-dmz-example-1/azuredeploy.json:
```diff
@@ -0,0 +1,521 @@
+    {
+      "type": "Microsoft.Storage/storageAccounts",
+      "name": "[parameters('newStorageAccountName')]",
+      "apiVersion": "2015-05-01-preview",
```
Use 2016-01-01 API version for storage. This is a breaking change, see 101-storage-account-create template for usage

singhkay commented on 411-dmz-example-1/azuredeploy.json:
```diff
@@ -0,0 +1,521 @@
+      }
+    },
+    {
+      "apiVersion": "2015-05-01-preview",
```
Use GA API version 2015-06-15

singhkay commented on 411-dmz-example-1/azuredeploy.json:
```diff
@@ -0,0 +1,521 @@
+﻿{
```
Reviewed first template but comments apply to all

singhkay commented on 411-dmz-example-1/README.md:
```diff
@@ -0,0 +1,10 @@
+# Build a DMZ with Network Security Groups (NSG)
```
A word on the folder naming for these, I would call these something along these lines that talks about the features in the folder name

* 300-dmz-nsg
* 300-dmz-nsg-firewall
* 300-dmz-nsg-firewall-udr

gatneil commented:
@tracsman any update on this?

tracsman commented:
Ok, finally getting back to this one. I've removed two templates to make the testing, review, and publish faster. I'll add the other two back in once I complete testing.

Based on feedback above, I've made the following changes:

- Removed storage name parameter and added a new variable:
    "StorageAccountName": "[concat(uniquestring(resourceGroup().id), 'dmz1')]",
- Also removed the parameter from the parameter json file.
- Updated all API versions from preview to current for each object.
- removing all but the first template from this pull request, will fix and re-add the other templates later.
- Renamed the folder from "411-dmz-example-1" to "300-dmz-nsg".
- Updated the metadata.json

I believe this covers all the feedback above.

tracsman commented:
After three more bug fixes, I think this template is ready to go, @gatneil what do you think?

singhkays commented:
@tracsman Can you call these templates 301 instead of 300? that makes it inconsistent with all the other templates

tracsman commented:
@singhkays Updated the folder name.

gatneil commented:
Could you take a quick look to make sure you're following the best practices checklist? https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

I think I noticed one issue from the checklist, which is the use of 'blob.core.windows.net' for vhd uri.

gatneil commented:
Any update on this?

tracsman commented:
@gatneil I've updated the blob references, how's it look now?
Thanks,
-Jon

gatneil commented on 301-dmz-nsg/README.md:
```diff
@@ -0,0 +1,10 @@
+# Build a DMZ with Network Security Groups (NSG)
+
+<a href="https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F300-dmz-nsg%2Fazuredeploy.json" target="_blank">
```
This url should have 301-dmz-nsg, not 300-dmz-nsg

gatneil commented on 301-dmz-nsg/README.md:
```diff
@@ -0,0 +1,10 @@
+<a href="https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F300-dmz-nsg%2Fazuredeploy.json" target="_blank">
+    <img src="http://azuredeploy.net/deploybutton.png"/>
+</a>
+<a href="http://armviz.io/#/?load=https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F300-dmz-nsg%2Fazuredeploy.json" target="_blank">
```
same

gatneil commented:
Noticed a small issue with links in the README. Otherwise looks good to go :)

gatneil commented:
Looks good to me; merging :)

tracsman commented:
@gatneil Good catch, totally forgot there were references in the ReadMe, they are fixed now.

tracsman commented:
@gatneil Thanks!!!
